### PR TITLE
Refactored load mapping file version 2.0

### PR DESCRIPTION
**Pull request by Linary**
Implement #35 #71

### CODE REVIEW

File: src/main/java/com/baidu/hugegraph/loader/executor/LoadContext.java [hunk 1]
```diff
@@ -81,7 +74,7 @@ public static synchronized void destroy() {
     }
 
     public static LoadContext get() {
-        E.checkState(instance != null, "Must init LoadContext firstly");
+        E.checkState(instance != null, "LoadContext Must be inited firstly");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/executor/LoadContext.java]: must be initialized

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 2]
```diff
@@ -50,66 +55,68 @@ public static void printRealtimeProgress() {
...
 
     private static void printMeterReport(LoadSummary summary) {
+        printAndLog("meter metrics");
         printAndLog("total time", TimeUtil.readableTime(summary.totalTime()));
-//        printAndLog("vertex parse rate", summary.parseRate(ElemType.VERTEX));
-//        printAndLog("vertex load rate", summary.loadRate(ElemType.VERTEX));
-//        printAndLog("edge parse rate", summary.loadRate(ElemType.EDGE));
-//        printAndLog("edge load rate", summary.loadRate(ElemType.EDGE));
+        printAndLog("vertex load rate", summary.loadRate(ElemType.VERTEX));
+        printAndLog("edge load rate", summary.loadRate(ElemType.EDGE));
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: add unit

File: src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java [hunk 3]
```diff
@@ -153,6 +155,15 @@ protected Object mappingFieldValueIfNeeded(String fieldName,
         return this.mapping().mappingValue(fieldName, fieldStrValue);
     }
 
+    protected boolean vertexIdEmpty(VertexLabel vertexLabel, Object vertexId) {
+        IdStrategy idStrategy = vertexLabel.idStrategy();
+        if (idStrategy.isCustomizeString()) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java]: same as primary id

Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java]: primary id will not be empty

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java [hunk 4]
```diff
@@ -178,6 +178,21 @@ public void shutdown() {
         }
     }
 
+    /**
+     * Execute parse task sync
+     */
+    public void executeParseTask(InputStruct struct, ElementMapping mapping,
+                                 ParseTaskBuilder.ParseTask task) {
+        List<Record> batch = task.get();
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: Do parse here?

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 5]
```diff
@@ -115,8 +115,8 @@ private static void printCountReport(LoadReport report) {
     private static void printMeterReport(LoadSummary summary) {
         printAndLog("meter metrics");
         printAndLog("total time", TimeUtil.readableTime(summary.totalTime()));
-        printAndLog("vertex load rate", summary.loadRate(ElemType.VERTEX));
-        printAndLog("edge load rate", summary.loadRate(ElemType.EDGE));
+        printAndLog("vertex load rate(rec/s)", summary.loadRate(ElemType.VERTEX));
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: prefer vertices/s

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 6]
```diff
@@ -115,8 +115,8 @@ private static void printCountReport(LoadReport report) {
...
+        printAndLog("edge load rate(rec/s)", summary.loadRate(ElemType.EDGE));
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: edges/s

File: src/main/java/com/baidu/hugegraph/loader/struct/EdgeStructV1.java [hunk 7]
```diff
@@ -68,6 +69,6 @@ public void check() throws IllegalArgumentException {
 
     @Override
     public String toString() {
-        return String.format("EdgeStruct(%s)", this.uniqueKey());
+        return String.format("EdgeMapping(%s)", this.uniqueKey());
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/struct/EdgeStructV1.java]: keep compatible EdgeStruct

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java [hunk 8]
```diff
@@ -143,37 +178,72 @@ public void shutdown() {
...
+            this.parseSemaphore.acquire();
+        } catch (InterruptedException e) {
+            throw new LoadException("Interrupted while waiting to submit %s " +
+                                    "parse task", e);
+        }
+
+        CompletableFuture.supplyAsync(task, this.parseService)
+                         .exceptionally(e -> {
+            LOG.error("Failed to execute parse task", e);
+            return null;
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: handle parse failures count

File: src/main/java/com/baidu/hugegraph/loader/util/MappingUtil.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            inputStructs.add(inputStruct);
+        }
+        for (InputStruct inputStruct : jdbcSourceInputStructs) {
+            inputStruct.id(String.valueOf(++id));
+            inputStructs.add(inputStruct);
+        }
+        return new LoadMapping(inputStructs);
+    }
+
+    private static ElementMapping convert(ElementStructV1 origin) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/MappingUtil.java]: convertV1ToV2

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 10]
```diff
@@ -22,162 +22,153 @@
...
+        LoadSummary summary = context.summary();
+        long vertexLoaded = summary.vertexLoaded();
+        long edgeLoaded = summary.edgeLoaded();
+        if (context.options().printProgress) {
+            Printer.printInBackward(vertexLoaded, edgeLoaded);
+        }
+
+        long loadSuccess = type.isVertex() ? vertexLoaded : edgeLoaded;
+        if (loadSuccess % frequency < batchSize) {
+            LOG.info("{} has been loaded: {}, with average load rate: {}/s",
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: remove "with"

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java (hunk 4 above)
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: yes

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java (hunk 8 above)
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: handled in ParseTaskBuilder.handleParseFailure(), actually there should be no exceptions here

File: src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java [hunk 11]
```diff
@@ -58,14 +59,19 @@ public void run() {
...
                     throw e;
                 }
+                if (StringUtils.containsAny(message, UNACCEPTABLE_MESSAGES)) {
+                    throw e;
+                }
                 retryCount = this.waitThenRetry(retryCount, e);
             }
         } while (retryCount > 0 && retryCount <= this.options().retryTimes);
 
+        // TODO：need to write to error log when don't addBatch succeed
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java]: when addBatch fails

File: src/main/java/com/baidu/hugegraph/loader/failure/FailWriter.java [hunk 12]
```diff
@@ -92,6 +104,25 @@ private void writeLine(String line) throws IOException {
         this.writer.newLine();
     }
 
+    private void writeHeaderIfNeeded() {
+        if (this.struct.input().header() == null || this.writedHeader) {
+            return;
+        }
+        synchronized(this.file) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/failure/FailWriter.java]: add space

File: src/main/java/com/baidu/hugegraph/loader/reader/jdbc/JDBCReader.java [hunk 13]
```diff
@@ -68,6 +69,7 @@ public void init(LoadContext context, InputStruct struct)
 
     @Override
     public void confirmOffset() {
+        // TODO: save offset
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/jdbc/JDBCReader.java]: still to be improved?

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java [hunk 14]
```diff
@@ -72,19 +67,13 @@ public TaskManager() {
          * limit the number of tasks added. When there are no idle threads in
          * the thread pool, the producer will be blocked, so OOM will not occur.
          */
-        this.parseService = ExecutorUtil.newFixedThreadPool(
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: why remove parseService

Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: At present, the code complexity is reduced first, and after a simple test, parallel parsing has only a slight performance improvement over serial

File: src/main/java/com/baidu/hugegraph/loader/reader/jdbc/JDBCReader.java (hunk 13 above)
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/reader/jdbc/JDBCReader.java]: Not required for this pr

File: src/main/java/com/baidu/hugegraph/loader/builder/EdgeBuilder.java [hunk 15]
```diff
@@ -59,32 +59,43 @@ public EdgeMapping mapping() {
...
-
-        Object targetVertexId = this.buildVertexId(this.targetLabel,
-                                                   this.mapping.targetFields(),
-                                                   properties);
-        if (this.vertexIdEmpty(this.targetLabel, targetVertexId)) {
-            return VIRTUAL_EDGE;
+        E.checkArgument(sources.size() == 1 || targets.size() == 1 ||
+                        sources.size() == targets.size(),
+                        "The source and target vertices must have the same " +
+                        "id number or one of them must be 1");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/builder/EdgeBuilder.java]: The elements number of source and target must be: 1 to n, n to 1, n to n.

File: src/test/java/com/baidu/hugegraph/loader/test/functional/FileLoadTest.java [hunk 16]
```diff
@@ -2031,6 +2029,7 @@ public void testLoadIncrementalModeAndLoadFailure()
         Assert.assertEquals(3, softwareFailureLines.size());
         Assert.assertEquals("lop,java,应该是数字", softwareFailureLines.get(2));
 
+        // TODO: 先只改一行,让第二行也出错
```
Comment (javeme) [inline src/test/java/com/baidu/hugegraph/loader/test/functional/FileLoadTest.java]: remove it

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 17]
```diff
@@ -280,6 +283,25 @@ private void handleReadFailure(InputStruct struct, ReadException e) {
...
+        LoadOptions options = this.context.options();
+        long failures = this.context.summary().totalParseFailures();
+        if (options.maxParseErrors != Constants.NO_LIMIT &&
+            failures >= options.maxParseErrors) {
+            if (this.context.stopped()) {
+                return;
+            }
+            synchronized (this.context) {
+                if (!this.context.stopped()) {
+                    Printer.printError("More than %s parsing error, stop " +
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: parse errors

File: src/main/java/com/baidu/hugegraph/loader/executor/LoadOptions.java [hunk 18]
```diff
@@ -202,12 +202,10 @@ public static LoadOptions parseOptions(String[] args) {
                         "The option --incremental-mode and --failure-mode " +
                         "can't be true at same time");
         if (options.failureMode) {
-            LOG.warn("The value of options: --max-read-errors, " +
-                     "--max-parse-errors and --max-insert-errors will be " +
-                     "setted as 1 in load-failure-mode");
-            options.maxReadErrors = 1;
-            options.maxParseErrors = 1;
-            options.maxInsertErrors = 1;
+            // NOTE: scan the entire error file in failure mode
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/executor/LoadOptions.java]: add LOG.info

File: src/main/java/com/baidu/hugegraph/loader/failure/FailWriter.java [hunk 19]
```diff
@@ -128,11 +140,28 @@ public void close() {
...
+
+    private void renameTempFile() {
+        // Renamed file if needed
+        boolean needRename = !this.options.incrementalMode;
+        if (needRename) {
+            String fileName = this.file.getAbsolutePath();
+            int idx = fileName.lastIndexOf(Constants.TEMP_FAILURE_SUFFIX);
+            String destName = fileName.substring(0, idx);
+            if (!this.file.renameTo(new File(destName))) {
+                LOG.warn("The failure file {} rename to {} failed",
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/failure/FailWriter.java]: Failed to rename failure file {} to {}

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 20]
```diff
@@ -232,16 +232,18 @@ private void load(InputStruct struct, InputReader reader) {
...
                 }
-                lines = new ArrayList<>(batchSize);
-
                 // Confirm offset to avoid lost records
                 reader.confirmOffset();
                 this.context.newProgress().markLoaded(struct, finished);
+
+                this.markStopIfNeeded();
+                if (this.context.stopped()) {
+                    LOG.warn("Parse errors exceed limit, load task stopped");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: stopped loading tasks

**Pull request merged**